x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java

```diff
@@ -226,50 +236,97 @@ public int get(int i) {
+
+        LoadFromMany(Block[] target, DocVector docVector) {
+            this.target = target;
+            shards = docVector.shards();
+            segments = docVector.segments();
+            docs = docVector.docs();
+            forwards = docVector.shardSegmentDocMapForwards();
+            backwards = docVector.shardSegmentDocMapBackwards();
+            builders = new Block.Builder[target.length];
+            rowStride = new BlockLoader.RowStrideReader[target.length];
+        }
+
+        void run() throws IOException {
+            for (int f = 0; f < fields.length; f++) {
+                /*
+                 * Important note: each block loader has a method to build an
+                 * optimized block loader, but we have *many* fields and some
+                 * of those block loaders may not be compatible with each other.
+                 * So! We take the least common denominator which is the loader
+                 * from the element expected element type.
+                 */
+                builders[f] = fields[f].info.type.newBlockBuilder(docs.getPositionCount(), blockFactory);
+            }
+            int p = forwards[0];
+            int lastShard = shards.getInt(p);
+            int lastSegment = segments.getInt(p);
+            int firstDoc = docs.getInt(p);
+            LeafReaderContext ctx = ctx(lastShard, lastSegment);
+            for (FieldWork field : fields) {
+                field.moveTo(lastShard, lastSegment, firstDoc);
             }
-            int lastShard = -1;
-            int lastSegment = -1;
-            BlockLoaderStoredFieldsFromLeafLoader storedFields = null;
-            for (int i = 0; i < forwards.length; i++) {
-                int p = forwards[i];
+            fieldsMoved(ctx, lastShard, lastSegment);
+            read(firstDoc);
+            for (int i = 1; i < forwards.length; i++) {
+                p = forwards[i];
                 int shard = shards.getInt(p);
                 int segment = segments.getInt(p);
-                int doc = docs.getInt(p);
                 if (shard != lastShard || segment != lastSegment) {
                     lastShard = shard;
                     lastSegment = segment;
-                    StoredFieldsSpec storedFieldsSpec = storedFieldsSpecForShard(shard);
-                    LeafReaderContext ctx = ctx(shard, segment);
-                    storedFields = new BlockLoaderStoredFieldsFromLeafLoader(
-                        StoredFieldLoader.fromSpec(storedFieldsSpec).getLoader(ctx, null),
-                        storedFieldsSpec.requiresSource() ? shardContexts.get(shard).newSourceLoader.get().leaf(ctx.reader(), null) : null
-                    );
-                    if (false == storedFieldsSpec.equals(StoredFieldsSpec.NO_REQUIREMENTS)) {
-                        trackStoredFields(storedFieldsSpec, false);
+                    for (FieldWork field : fields) {
+                        field.clearAndMoveTo(shard, segment);
```

Comment:
It appears that we are not reloading the `FieldWork.loader` when the shard changes with `clearAndMoveTo`?

Reply:
Oh darn. Let me look.That's bad. We totally should.

Reply:
OK! I've reworked things a little and I think they are better.

Reply:
Yes, the new method `positionFieldWork` is better.